### DATA MANAGEMENT / 4

This i a base script to concatenate differents parts of satellite data... This step takes place after the data collection and aggreagation...
First, processing on sattelite images probably take several steps (with differents parts for differents problematics plots) and we need to assembly theses.
Then, we have 4 differents xlsx files (for our 4 local machines...) and we just want to build one file, merged with others data types (IFN, meteo...)

In [1]:
import pandas as pd
import numpy as np
import copy

FIRST USAGE / CONCENATION OF DIFFERENTS PARTS OF A DATAFRAME WITH SAME LFI :

In [2]:
nb_parts = 3 #nb dataframes you have to join

LFI_name = 'LFI1'

In [3]:
dataframes_results =[]
dataframes_results.append(pd.read_excel('DATA_SAT_aggregated/Data_from_satellites_images_LFI1.xlsx').drop('Unnamed: 0', axis=1))
dataframes_results.append(pd.read_excel('DATA_SAT_aggregated/Data_from_satellites_images_LFI1_pb1.xlsx').drop('Unnamed: 0', axis=1))
dataframes_results.append(pd.read_excel('DATA_SAT_aggregated/Data_from_satellites_images_LFI1_pb2.xlsx').drop('Unnamed: 0', axis=1))

In [4]:
df_result_base = dataframes_results[0][['PARCELLE','LFI']]

In [5]:
dataframes_results[0]

PARCELLE   LFI    NDVI     EVI    NDMI    NDWI    DSWI
0           51  LFI1  0.3779  0.0146  0.2264 -0.3325  0.4705
1          384  LFI1  0.2861  0.0071  0.3293 -0.2327  0.3518
2         1239  LFI1  0.6749  0.0219  0.2021 -0.5865  0.8626
3         1419  LFI1  0.6092  0.0201  0.1112 -0.5143  0.6521
4         1431  LFI1  0.7039  0.0245  0.1949 -0.5983  0.8834
...        ...   ...     ...     ...     ...     ...     ...
2398    164918  LFI1  0.6280  0.0208  0.1838 -0.5329  0.7604
2399    164922  LFI1  0.6796  0.0253  0.2015 -0.5812  0.8639
2400    164999  LFI1  0.6270  0.0211  0.1165 -0.5465  0.6925
2401    165003  LFI1  0.6385  0.0184  0.2485 -0.5291  0.8416
2402    165065  LFI1  0.6815  0.0224  0.2499 -0.5828  0.9330

[2403 rows x 7 columns]

All the parts of a LFI campaign file have 2403 rows (it's due to the automation of the script wich have maked the aggregation of each list of satellite data...) So, we have a problem to assemble them, because we have to write a new value taken from a second file just if we have a empty value in the first file...

We need to build a loop to do this:

In [6]:
nb_lines = len(df_result_base)
result_arr = np.empty((nb_lines,5)) # we create an empty numpy matrix to stock all results

for i in range(nb_parts):
    c=0
    arr = np.empty((nb_lines,5))
    arr = dataframes_results[i].iloc[:,2:].to_numpy()
    for line in range(nb_lines):
        for col in range(5):
            if str(arr[line][col]) != 'nan':
                result_arr[line][col] = copy.copy(arr[line][col])
                c+=1
    print(f'Join n°{i} : {c} data copied ...')

Join n°0 : 11785 data copied ...
Join n°1 : 210 data copied ...
Join n°2 : 20 data copied ...


Conversion of the results in a dataframe :

In [7]:
df_result = copy.copy(df_result_base)
df_result[['NDVI','EVI','NDMI','NDWI','DSWI']] = result_arr
df_result

PARCELLE   LFI    NDVI     EVI    NDMI    NDWI    DSWI
0           51  LFI1  0.3779  0.0146  0.2264 -0.3325  0.4705
1          384  LFI1  0.2861  0.0071  0.3293 -0.2327  0.3518
2         1239  LFI1  0.6749  0.0219  0.2021 -0.5865  0.8626
3         1419  LFI1  0.6092  0.0201  0.1112 -0.5143  0.6521
4         1431  LFI1  0.7039  0.0245  0.1949 -0.5983  0.8834
...        ...   ...     ...     ...     ...     ...     ...
2398    164918  LFI1  0.6280  0.0208  0.1838 -0.5329  0.7604
2399    164922  LFI1  0.6796  0.0253  0.2015 -0.5812  0.8639
2400    164999  LFI1  0.6270  0.0211  0.1165 -0.5465  0.6925
2401    165003  LFI1  0.6385  0.0184  0.2485 -0.5291  0.8416
2402    165065  LFI1  0.6815  0.0224  0.2499 -0.5828  0.9330

[2403 rows x 7 columns]

Missing values ?

In [8]:
df_result['NDVI'].isnull().sum()

0

It's ok

We can export it :

In [9]:
df_result.to_excel(f'./DATA_SAT_aggregated/Data_from_satellites_images_{LFI_name}_total.xlsx')

--------------

SECOND USAGE / CONCATENATION OF ALL DATAFRAMES FROM DIFFERENTS LFI :

After assembling all the files for each campaign, we need to assemble all the files, from our 4 local machines... It's easier...

In [15]:
data_1 = pd.read_excel('DATA_SAT_aggregated/Data_from_satellites_images_LFI1.xlsx').drop('Unnamed: 0', axis=1)
data_2 = pd.read_excel('DATA_SAT_aggregated/Data_from_satellites_images_LFI2.xlsx').drop('Unnamed: 0', axis=1)
data_3 = pd.read_excel('DATA_SAT_aggregated/Data_from_satellites_images_LFI3.xlsx').drop('Unnamed: 0', axis=1)
data_4 = pd.read_excel('DATA_SAT_aggregated/Data_from_satellites_images_LFI4.xlsx').drop('Unnamed: 0', axis=1)

In [18]:
data_concate = pd.concat([data_1, data_2, data_3, data_4], axis=0)

In [19]:
data_concate

PARCELLE   LFI    NDVI     EVI    NDMI    NDWI    DSWI
0           51  LFI1  0.3779  0.0146  0.2264 -0.3325  0.4705
1          384  LFI1  0.2861  0.0071  0.3293 -0.2327  0.3518
2         1239  LFI1  0.6749  0.0219  0.2021 -0.5865  0.8626
3         1419  LFI1  0.6092  0.0201  0.1112 -0.5143  0.6521
4         1431  LFI1  0.7039  0.0245  0.1949 -0.5983  0.8834
...        ...   ...     ...     ...     ...     ...     ...
2398    164918  LFI4  0.5786  0.0179  0.3073 -0.4821  0.8203
2399    164922  LFI4  0.4165  0.0162  0.2601 -0.3596  0.5811
2400    164999  LFI4  0.3737  0.0144  0.0118 -0.2744  0.3005
2401    165003  LFI4 -0.0614 -0.0075  0.1396  0.0614 -0.0693
2402    165065  LFI4 -0.1602 -0.0119  0.0199  0.1521 -0.1549

[9612 rows x 7 columns]

In [24]:
big_data = pd.read_excel('DATA_preprocess_EDA_base/big_merge_V2_meteo.xlsx', sheet_name='Sheet1').drop('Unnamed: 0', axis=1)

In [25]:
big_data = big_data.merge(data_concate, how='outer', on=['PARCELLE','LFI'])

Export :

In [27]:
big_data.to_excel('DATA_preprocess_EDA_base/big_merge_V2_meteo_SAT.xlsx')